feat(upgrade): additional ng2 features for downgraded ng2 components

youdz opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

Note: Documentation doesn't mention the absence of support for the added features, the user is lead to believe they are implicitly supported. So no changes on the [upgrade tutorial](https://angular.io/docs/ts/latest/guide/upgrade.html) is needed.

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Angular 2 components downgraded with the UpgradeAdapter are missing several critical features, like multi-slot projection and parent component/providers injection.

**What is the new behavior?**
These features are &now supported by the UpgradeAdapter.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

**Other information**:
- Full use of core Angular 2 projection for downgraded Angular 2
  components. In particular, this enables multi-slot projection and
  other features on &lt;ng-content&gt;.
- Hierarchical injectors for downgraded Angular 2 components: downgraded
  components inherit the injector of the first other downgraded Angular
  2 component they find up the DOM tree.
- declareNg1Module() and initForNg1Tests() methods on the UpgradeAdapter
  to allow testing hybrid applications through Angular 1 without having
  to redeclare the adapter module and recompile every downgraded component
  for every test.

Closes #6629, #7727, #8729, #5462

youdz committed:
feat(upgrade): additional ng2 features for downgraded ng2 components

- Full use of core Angular 2 projection for downgraded Angular 2
  components. In particular, this enables multi-slot projection and
  other features on <ng-content>.
- Hierarchical injectors for downgraded Angular 2 components: downgraded
  components inherit the injector of the first other downgraded Angular
  2 component they find up the DOM tree.
- declareNg1Module() and initForNg1Tests() methods on the UpgradeAdapter
  to allow testing hybrid applications through Angular 1 without having
  to redeclare the adapter module and recompile every downgraded component
  for every test.

Closes #6629, #7727, #8729, #5462, #9643, #9649

vicb commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -226,74 +241,52 @@ export class UpgradeAdapter {
             .resolveAndCreate(
                 [
                   BROWSER_APP_PROVIDERS, BROWSER_APP_COMPILER_PROVIDERS,
-                  {provide: NG1_INJECTOR, useFactory: () => ng1Injector},
```
why ? `provide()` is deprecated

youdz replied:
Old rebase gone wrong, sorry about that.
Fixing it right now.

vicb commented:
Could you please `gulp lint` & update ?

youdz commented:
@vicb: removed provide(), added the missing semicolon, and updated the review.

vicb commented:
@mhevery could you please review ?

vicb commented:
Please update the public API test (run locally with ./test.sh tools).

Breaking changers must be reported in the commit comment:
```
BREAKING CHANGE:

Description

Before:
    
    code

After:

    code
```

Before/After only if required.

Please also squash your commits

youdz commented:
I updated the public API tests before rebasing, then figured I might as well rebase my PR now if I'm updating it. `./test.sh tools` give me "No specs found" on the current HEAD.

Also, as checked in the initial comment, this PR does not introduce any breaking changes.

youdz commented:
So e2e tests still fail. Why do imports from `@angular/xxx/src/` fail in these? I'd be happy to fix it if you could point me in the right direction. To be honest, I'm struggling a bit to find documentation on the current build and tests, even [DEVELOPER.md](DEVELOPER.md) is obsolete.

youdz commented:
There, all checks are green. I tried to follow the patterns I found regarding private exports.

vicb commented:
You can only import from top level in app (hence e2e), ie `@angular/core`, `@angular/common, ...

mhevery commented on modules/@angular/upgrade/src/facade:
```diff
@@ -0,0 +1 @@
+../../facade/src
```
ngUpgrade should not depend on Facade. Please remove.

mhevery commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -6,13 +6,17 @@
-import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INJECTOR, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
+import {NG1_COMPILE, NG1_INJECTOR, NG1_PARSE, NG1_ROOT_SCOPE, NG1_TESTABILITY, NG2_COMPILER, NG2_COMPONENT_FACTORY_REF_MAP, NG2_INIT_PROMISE_COMPLETER, NG2_INJECTOR, NG2_RESOLVED_METADATA_MAP, NG2_ZONE, REQUIRE_INJECTOR} from './constants';
 import {DowngradeNg2ComponentAdapter} from './downgrade_ng2_adapter';
+import {isBlank, isPromise} from './facade/lang';
```
remove facade.

mhevery commented:
LGTM, but I would like to remove the facade as it is not needed.

youdz commented:
I removed the facade, just added a PromiseCompleter to `util.ts` because we need one anyway. But rebasing on the latest master broke almost every UpgradeAdapter test.

Trying to figure out what is happening exactly what is happening, I believe is that the NgZone is not stable yet when `ready()` is called. I'll update the PR once the rebase is through and all tests pass.

youdz commented:
@mhevery 
- facade removed
- rebased on latest master
- all tests green

I'm still open to suggestions if you see anything else that should change.

youdz commented:
Some people at VMware have been trying this UpgradeAdapter, and I have been told that Angular 1's `$compile` asynchronous (as we do by chaining injector promises in linking functions) poses a lot of problems, both during testing and on flickering views in application.

So I'll change it to synchronous promise-like objects and update the review, but I wanted to warn you since it'll introduce a change you haven't checked out.

And of course, if you think making Angular 1's linking process of downgraded Angular 2 components is a bad idea, I'd be curious to know why.

I apologize for the extra review step.

youdz commented:
Rebased on the latest master to solve conflicts. Let me post a couple comments on specific points.

youdz commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -94,7 +94,7 @@ export function main() {
            adapter.bootstrap(element, ['ng1']).ready((ref) => {
              expect(document.body.textContent).toEqual('1A;2A;ng1a;2B;ng1b;2C;1C;');
              // https://github.com/angular/angular.js/issues/12983
-             expect(log).toEqual(['1A', '1B', '1C', '2A', '2B', '2C', 'ng1a', 'ng1b']);
```
@mhevery (because you're the one who introduced this test, I think): I believe the ideal behavior is without the `'1B'`. The `ng2` directive doesn't project its content anywhere, it just replaces it by its own template, so it shouldn't be linked. 
This is already the case in Angular 1: http://plnkr.co/edit/ILUdsFCyJWvNR34b6c6a?p=preview
The log only shows "A C", the B is never linked.

If I missed a good reason to link content that will be discarded anyway, I'm happy to revert this and try to have downgraded components respect this behavior.

youdz commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -291,20 +278,27 @@ export class UpgradeAdapter {
+    var ng2Injector = this.ng2Injector = this.applicationRef.injector;
+
+    let compilerInjector =
+        ReflectiveInjector.resolveAndCreate([BROWSER_APP_COMPILER_PROVIDERS], platformRef.injector);
```
`BROWSER_APP_COMPILER_PROVIDERS` is deprecated since the switch to `@AppModule`.
I understand the compiler is becoming entirely private, which makes complete sense. But the UpgradeAdapter needs access to a `CompileMetadataResolver` and a `DirectiveNormalizer`, which are now "hidden" from the application. The compiler uses its own private injector.

What would you recommend to access these compiler objects here ? Thanks.

kseamon replied:
Sort of a hack, but perhaps you could use compilerOptions (see #10575) to inject a special provider into the compiler that could then pass back the needed parts? Not sure if that's kosher though, @mhevery?

kseamon commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -528,39 +674,109 @@ export class UpgradeAdapter {
+            if (!componentFactory)
+              throw new Error('Expecting ComponentFactory for: ' + info.selector);
+
+            element.empty();
```
Can this be done in the compile step?

kseamon commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -528,39 +674,109 @@ export class UpgradeAdapter {
+        return {
+          post: (scope: angular.IScope, element: angular.IAugmentedJQuery,
+                 attrs: angular.IAttributes, parentInjector: Injector | ParentInjectorPromise,
+                 transclude: angular.ITranscludeFunction): void => {
```
The transculde params are not used here or in the compile method above and can be removed.

kseamon commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -528,39 +674,109 @@ export class UpgradeAdapter {
+
+            element.empty();
+            let projectableNodes = linkFns.map(link => {
+              let projectedClone: Node[];
```
Nit: IAugmentedJQuery would be more accurate (Here and a few other places)

kseamon commented on modules/@angular/upgrade/src/util.ts:
```diff
@@ -22,3 +27,62 @@ export function onError(e: any) {
+    }
+  }
+  return asArray;
+}
```
Performance nits:

var asArray: string[][];
 if (isPresent(attributes)) {
    let attrLen = attributes.length;
    asArray = new Array(attrLen);
    for (var i = 0; i < attrLen; i++) {
      asArray[i] = [attributes[i].nodeName, attributes[i].nodeValue];
    }
  }
  return asArray || [];

kseamon commented on modules/@angular/upgrade/src/util.ts:
```diff
@@ -22,3 +27,62 @@ export function onError(e: any) {
+    let selectors = resolvedMetadata.template.ngContentSelectors;
+    let matcher = new SelectorMatcher();
+    let wildcardNgContentIndex: number;
+    for (let i = 0; i < selectors.length; i++) {
```
similarly here,

for (let i = 0, ii = selectors.length; i < ii; i++) {

icfantv commented:
@youdz what else do you need for this?  what are your thoughts on @kseamon's requested changes?  @mhevery, this PR is currently blocking us at work, what else would you like to see to get this merged?

youdz commented:
Sorry it took me so long, everyone. Initially I rebased frequently on each Release Candidate, hoping this would merged in. But at some point I just gave up and decided to wait on Angular final.
With all the breaking changes accumulated, rebasing on the latest needed a major overhaul, so most patterns have changed. Thanks @kseamon for your suggestion of using CompilerOptions, it put me on the right track.

The current solution has a hackier integration with the compiler, but the good news is that we don't re-normalize all the templates all the time. We just use the ones from the compiler.

Regarding the loose types @kseamon mentioned (Node instead of IAugmentedJQuery and others), the problem is that I used the strictest types I could that would respect the ones defined in `angular_js.ts`. Changing them is a major cleanup that impacts far more than the few parts I'm modifying. I'd rather get this through and clean up types across the entire upgrade module later than wait even longer on this. I hope that's fine.

@mhevery: Any chance we could merge this before the next release? The PR has been waiting for quite some time, and quite a few people seem to depend on it. Thanks. :-)

mhevery commented:
@youdz Yes the PR has a LGTM which means that it should be merged. But it also does not merge cleanly, which is the reason why it is not mergable and it is blocking the process. Could you rebase to make sure that we can merge it?

youdz commented:
I actually did rebase several times when I saw conflicts introduced in the master branch, but these are so frequent that I just waited for a signal to pick it up again. :)

I'll rebase again tomorrow, so hopefully you should have a mergeable version of this PR before the end of day tomorrow.

Thanks :)

petebacondarwin commented on modules/@angular/upgrade/src/downgrade_ng2_adapter.ts:
```diff
@@ -23,26 +23,23 @@ export class DowngradeNg2ComponentAdapter {
 
   constructor(
-      private id: string, private info: ComponentInfo, private element: angular.IAugmentedJQuery,
+      private info: ComponentInfo, private element: angular.IAugmentedJQuery,
```
Oh, maybe it doesn't appear in the public API... or at least it doesn't appear in the public_api_guard file.

petebacondarwin commented on modules/@angular/upgrade/src/platform-upgrade.ts:
```diff
@@ -0,0 +1,24 @@
+ * @experimental
+ */
+export const platformUpgrade =
+    createPlatformFactory(platformBrowserDynamic, 'upgrade', INTERNAL_UPGRADE_PLATFORM_PROVIDERS);
```
Oh, maybe it doesn't appear in the public API... or at least it doesn't appear in the public_api_guard file.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -269,67 +282,36 @@ export class UpgradeAdapter {
     var rootScope: angular.IRootScopeService;
     var componentFactoryRefMap: ComponentFactoryRefMap = {};
+    var compilerSpy: RuntimeCompilerSpy = platformUpgrade().injector.get(NG2_COMPILER_SPY);
+    var uprgadeAdapter = this;
```
typo: `uprgadeAdapter` -> `upgradeAdapter`

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -375,60 +357,200 @@ export class UpgradeAdapter {
+   *   };
+   * });
+   *
+   *
```
`var` -> `const`?

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -375,60 +357,200 @@ export class UpgradeAdapter {
+   * })
+   * class Ng2 {
+   * }
+   *
```
`let` -> `const`?

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_adapter.ts:
```diff
@@ -375,60 +357,200 @@ export class UpgradeAdapter {
+   * class Ng2 {
+   * }
+   *
+   * @NgModule({
```
`let` -> `const`?

petebacondarwin commented on modules/@angular/compiler/src/template_parser/template_parser.ts:
```diff
@@ -1162,7 +1162,8 @@ class ElementContext {
 
-function createElementCssSelector(elementName: string, matchableAttrs: string[][]): CssSelector {
+export function createElementCssSelector(
+    elementName: string, matchableAttrs: string[][]): CssSelector {
```
a change like this would be cleaner in a separate `style` commit

petebacondarwin replied:
Oops, sorry, I just realised that the actual change is to make it exported.

petebacondarwin commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -96,6 +97,8 @@ export interface IAugmentedJQuery {
   contents(): IAugmentedJQuery;
   parent(): IAugmentedJQuery;
+  empty(): void;
+  append(nodes: Node[]): IAugmentedJQuery;
```
these two lines are already in master

petebacondarwin commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -30,6 +30,7 @@ export interface IRootScopeService {
   $new(isolate?: boolean): IScope;
   $id: string;
   $watch(expr: any, fn?: (a1?: any, a2?: any) => void): Function;
+  $on(event: string, fn?: (event?: any, ...args: any[]) => void): Function;
```
this line is already in master

petebacondarwin commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -133,7 +136,7 @@ function noNg() {
 
 var angular: {
-  bootstrap: (e: Element, modules: string[], config: IAngularBootstrapConfig) => void,
+  bootstrap: (e: Element, modules: (string | any[])[], config: IAngularBootstrapConfig) => void,
```
this has been fixed already in master

petebacondarwin commented on modules/@angular/upgrade/src/spies.ts:
```diff
@@ -0,0 +1,50 @@
+
+  add(template: any) {
+    if (!template.isHost) {
+      let metadata = template.normalizedCompMeta;
```
`let` -> `const`

petebacondarwin commented on modules/@angular/upgrade/src/spies.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+@Injectable()
+export class RuntimeCompilerSpy {
```
I would find this more grokkable if it was simply called something like `CapturedContentSelectors` and the `add` method was inside the `RuntimeCompilerSpyFactory`

petebacondarwin commented on modules/@angular/compiler/src/private_export.ts:
```diff
@@ -104,7 +114,10 @@ export const __compiler_private__: {
+  RuntimeCompilerFactory: compiler.RuntimeCompilerFactory,
   StyleCompiler: style_compiler.StyleCompiler,
   ViewCompiler: view_compiler.ViewCompiler,
   NgModuleCompiler: ng_module_compiler.NgModuleCompiler,
```
This file is no longer needed https://github.com/angular/angular/commit/acda82c1edf01acffcf208ddc10622d1aa8fb4e1

petebacondarwin commented:
I am rebasing and breaking this up into smaller commits.

youdz commented:
@petebacondarwin: I saw your comments after I was done rebasing, so I figured I might as well update the PR. Maybe it will help.

I took your comments into account, thank you for your help.

petebacondarwin commented:
Thanks @youdz. Just to keep you updated. I will have a branch for you to review within an hour.

vicb commented:
What is the status here ?

youdz commented:
Still waiting on #12675 to go in. Not sure what you want to do with this pull request while it's sorted out (which shouldn't take long, it's been approved several times now).

petebacondarwin commented:
Hurray!  #12675 has finally landed. Thanks @youdz for your patience.
